[llvm] Change `fp128` lowering to use `f128` functions by default

tgross35 opened the pull request:
LLVM currently emits calls to `*l` (`long double`) libm symbols for `fp128` intrinsics. This works on platforms where `long double` and `_Float128` are the same type, but is incorrect on many platforms.

Change RuntimeLibcalls such that `*f128` libcalls are used by default, which is always safe and correct but may not be available. On platforms where `long double` is `f128`, the current lowering to `*l` symbols is kept since they are more available.

The logic for whether f128 is `long double` is based on the platforms in Clang that set `LongDoubleFormat` to `llvm::APFloat::IEEEquad`.

Fixes: https://github.com/llvm/llvm-project/issues/44744
Discourse discussion: https://discourse.llvm.org/t/fp128-math-functions-strange-results/72708
Initial patchset: https://reviews.llvm.org/D157836

Includes PRs:
* https://github.com/llvm/llvm-project/pull/148349
* https://github.com/llvm/llvm-project/pull/148308

tgross35 commented:
@efriedma-quic was looking at this on phabricator

efriedma-quic commented:
This is basically the approach I was expecting: we check the type of "long double" when we build the TargetLowering, and pick appropriate names based on that.

I expect that for -mlong-double-128, you just want to add a module flag that overrides the default choice.

I think I'd prefer to keep the clang type information computation independent from the backend's type information, even if it overlaps.  We try to layer the clang frontend so it isn't directly tied to LLVM IR outside of CodeGen.

My first thought was that the computation of the defaults should be in the backend, not Triple.h, since nothing else needs it at the moment.  But I guess it could be useful outside the backend, so maybe that's fine.  (At the moment, all the relevant optimizations just check the type of the call itself, but I can imagine certain optimizations could benefit from being able to compute the type without an existing signature to consult.)

tgross35 commented:
I'm struggling a bit with how to handle ABI information since that affects layout (e.g. ARM aapcs), which I think explains most of the errors in https://buildkite.com/llvm-project/github-pull-requests/builds/31198#018d26e2-fd17-4e15-a1eb-08580c189056. This needs to be available at TargetLoweringBase::InitLibcalls, which calls [`getCLayouts`](https://github.com/llvm/llvm-project/blob/cb3bf7540cf9b797575c625318e47a33f7514fad/llvm/lib/TargetParser/Triple.cpp#L1945).

TargetMachine is available at that time, so would it be better to move CLayouts from Triple to TargetMachine? If so subclasses could be used rather than the if block, which more closely follows the Clang side.

Also, are there currently any module flags that make it to TargetLowering? Looking for a reference on how get the -mlong-double-128 information.

efriedma-quic commented:
Putting a function in TargetMachine seems reasonable.

efriedma-quic commented:
For the question about querying module flags, we do that in a few different places in codegen; grep for "getModuleFlag".  Not sure if there's anything specifically in TargetLowering.

tgross35 commented:
Finally getting around to this after more than a year. @efriedma-quic as an alternative to the current implementation of duplicating `long double` layout information from Clang to LLVM, would it work if LLVM lowers to `*f128` calls but provides a module flag `fp128_use_long_double_libcalls` to prefer the `*l` versions? So if Clang or other frontends know that their `long double` is `_Float128`, it can select those libcalls.

The advantage is avoided code duplication and the logic is easier to follow. Also this avoids problems if linking a library built with an unexpected `-mlong-double-` configuration.

The disadvantage is that frontends that don't know about C's `long double` can't benefit from the more common `*l` symbols. I don't think this is too big of a problem though: it makes no difference with glibc (the f128 aliases have been around sufficiently long) or on any platforms where `long double` is not `_Float128`. And it is easy enough for frontends to set `fp128_use_long_double_libcalls` on a case-by-case basis if they know what math library is being used (e.g. aarch64 musl).

(I handle the f128 support for Rust and would much rather never think about `*l` symbols, I can alias them to `*f128` if needed or set the flag)

tgross35 commented:
In either case, I need to have the module flags available pretty early and I'm not sure how to do that. Ideally they would be available when `TargetLowering` is constructed or sometime before it is used for lowering, but it only gets a `TargetMachine` as a paremeter. All values in `TargetOptions` seem to be configured once and don't pay attention to module flags or take the module as a parameter - is there a reason for that? I'm wondering if `TargetMachine` is intended to be unchanging across different modules.

tgross35 commented on llvm/test/CodeGen/Generic/f128-math-lowering.ll:
```diff
@@ -0,0 +1,331 @@
+; RUN: llc < %s -mtriple=x86_64-unknown-linux-gnu -verify-machineinstrs | FileCheck %s --check-prefix=CHECK-NOTLD
+; RUN: llc < %s -mtriple=x86_64-unknown-linux-musl -verify-machineinstrs | FileCheck %s --check-prefix=CHECK-NOTLD
+;
+; REQUIRES: aarch64-registered-target
+; REQUIRES: powerpc-registered-target
+; REQUIRES: riscv-registered-target
+; REQUIRES: systemz-registered-target
+; REQUIRES: x86-registered-target
```
Todo: replace this with `%if` somehow so this test still runs if only a subset of architectures is available https://llvm.org/docs/TestingGuide.html

tgross35 commented:
Talked to arsenm on discord, long discussion starting around here https://discord.com/channels/636084430946959380/636732535434510338/1362207130559578185. The outcome is that this is effectively a target option and needs to be tied to the triple rather than per-module. Which makes sense and avoids the above problem.

So, I'll be doing the following:

1. Make LLVM assume that `sqrtf128` (and similar) libcalls are available by default
2. On 64-bit arm, loongarch, mips, risc-v, and s390x musl targets, use `sqrtl` instead
3. Add some way to make musl targets also use `sqrtf128`, like `-nolongdouble` in the target triple

This should work because calling `sqrtf128` is correct on _most_ platforms:
* On Windows, Apple, and 32-bit platforms, `long double` is `f64` so `sqrtf128` is the only correct call
* On x86, `long double` is the x87 80-bit float so `sqrtf128` is the only correct call
* On anything glibc, `sqrtf128` is an alias to `sqrtl` on platforms where that works, so `sqrtf128` can always be called
* That leaves 64-bit `musl` on platforms where `long double` is `f128` as the only platforms where `sqrtl` has to be called (otherwise calling `sqrtl` from C would get intercepted and relowered as `sqrtf128` for a linker error)

tgross35 commented on llvm/include/llvm/Analysis/TargetTransformInfoImpl.h:
```diff
@@ -189,27 +189,28 @@ class TargetTransformInfoImplBase {
-        Name == "tanh"  || Name == "tanhf"  || Name == "tanhl" ||
-        Name == "sqrt"  || Name == "sqrtf"  || Name == "sqrtl" ||
-        Name == "exp10"  || Name == "exp10l"  || Name == "exp10f")
+    if (Name == "copysign" || Name == "copysignf" || Name == "copysignl" || Name == "copysignf128" ||
+        Name == "fabs"  || Name == "fabsf"  || Name == "fabsl"  || Name  == "fabsf128" ||
+        Name == "fmin"  || Name == "fminf"  || Name == "fminl"  || Name  == "fminf128" ||
+        Name == "fmax"  || Name == "fmaxf"  || Name == "fmaxl"  || Name  == "fmaxf128" ||
+        Name == "sin"   || Name == "sinf"   || Name == "sinl"   || Name  == "sinf128"  ||
+        Name == "cos"   || Name == "cosf"   || Name == "cosl"   || Name  == "cosf128"  ||
+        Name == "tan"   || Name == "tanf"   || Name == "tanl"   || Name  == "tanf128"  ||
+        Name == "asin"  || Name == "asinf"  || Name == "asinl"  || Name  == "asinf128" ||
+        Name == "acos"  || Name == "acosf"  || Name == "acosl"  || Name  == "acosf128" ||
+        Name == "atan"  || Name == "atanf"  || Name == "atanl"  || Name  == "atanf128" ||
+        Name == "atan2" || Name == "atan2f" || Name == "atan2l" || Name  == "atan2f128"||
+        Name == "sinh"  || Name == "sinhf"  || Name == "sinhl"  || Name  == "sinhf128" ||
+        Name == "cosh"  || Name == "coshf"  || Name == "coshl"  || Name  == "coshf128" ||
+        Name == "tanh"  || Name == "tanhf"  || Name == "tanhl"  || Name  == "tanhf128" ||
+        Name == "sqrt"  || Name == "sqrtf"  || Name == "sqrtl"  || Name  == "sqrtf128" ||
+        Name == "exp10" || Name == "exp10l" || Name == "exp10f" || Name  == "exp10f128")
```
I'm not sure what the effects of this change are, but it seemed unusual that long double and f128 would be treated differently here.

tgross35 commented on llvm/test/CodeGen/PowerPC/f128-arith.ll:
```diff
@@ -425,14 +425,19 @@ define fp128 @qp_sincos(ptr nocapture readonly %a) nounwind {
+; CHECK-NEXT:    stxv v31, 48(r1) # 16-byte Folded Spill
+; CHECK-NEXT:    stxv v30, 32(r1) # 16-byte Folded Spill
+; CHECK-NEXT:    lxv v31, 0(r3)
+; CHECK-NEXT:    vmr v2, v31
+; CHECK-NEXT:    bl cosf128
+; CHECK-NEXT:    nop
+; CHECK-NEXT:    vmr v30, v2
+; CHECK-NEXT:    vmr v2, v31
+; CHECK-NEXT:    bl sinf128
 ; CHECK-NEXT:    nop
-; CHECK-NEXT:    lxv v2, 48(r1)
-; CHECK-NEXT:    lxv v3, 32(r1)
-; CHECK-NEXT:    xsmulqp v2, v3, v2
+; CHECK-NEXT:    xsmulqp v2, v30, v2
+; CHECK-NEXT:    lxv v31, 48(r1) # 16-byte Folded Reload
+; CHECK-NEXT:    lxv v30, 32(r1) # 16-byte Folded Reload
```
I don't follow this, why was the sin+cos->sincos optimization lost here?

tgross35 commented:
@arsenm would you mind taking a look at this? It should fix most of the problems with unsoundly lowering f128 to long double calls. No control via target string, I'll do that separately.

Does https://github.com/llvm/llvm-project/blob/50aacb9e1b51811f12cb017719678962411c50dd/llvm/lib/CodeGen/IntrinsicLowering.cpp#L386-L445 need to be updated somehow?

uweigand commented on llvm/test/CodeGen/SystemZ/fp-libcall.ll:
```diff
@@ -39,7 +39,7 @@ define double @f5(double %x, double %y) {
 define fp128 @f6(fp128 %x, fp128 %y) {
 ; CHECK-LABEL: f6:
-; CHECK: brasl %r14, powl@PLT
+; CHECK: brasl %r14, powf128@PLT
```
Why are we seeing all these changes in the SystemZ tests?   On SystemZ we should continue to use the *l names (I thought this is what the `shouldLowerf128AsLongDouble` logic was supposed to achieve?)

tgross35 replied:
The logic I have with that function is that if available, *f128 should be preferred over *l if because that is always correct for `fp128`. *l would be incorrect in the unusual case that the math library was compiled with a different `-mlong-double` flag. On s390x (and others where l-d is f128) the *f128 versions are only used on glibc targets, which this test is showing, because glibc aliases the *f128 and *l functions when it knows `long double` is binary128.

The downside of course is that compiling for glibc but bringing your own libm will need to add aliases as well. This seems acceptable to me because it should be the library's responsibility to indicate that it expects `long double` to be the same as `_Float128` rather than anything else. (Also per @arsenm on Discord LLVM assumes that the libm from your libc is used, or at least has the same interfaces).

If you're fine with the change then I can update this tests showing that a musl target is unaffected. Or if you would prefer I can drop the `isGNUEnvironment` check and always use *l on s390x, since it's less likely to be a problem here.

uweigand replied:
I'm a bit concerned about making this change by default, I don't see that this has any actual advantages on s390x (*l is always correct here anyway), and it might cause some surprises.  For example, while current glibc version do indeed provide the *f128 aliases, that has not always been the case - e.g. RHEL 7 doesn't have them.

tgross35 replied:
This has been updated to not change the current behavior on existing platforms, with a small exception of MIPS32 (due to lacking ABI information)

arsenm commented on clang/lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -394,6 +395,17 @@ static void checkDataLayoutConsistency(const TargetInfo &Target,
+      &Target.getLongDoubleFormat() != &llvm::APFloat::IEEEquad()) {
+    const char *SemName =
+        llvm::APFloatBase::SemanticsName(Target.getLongDoubleFormat());
+    llvm::errs() << "For target `" << Triple.str()
+                 << "` LLVM wants to use `long double` symbols for `_Float128` "
+                    "libm call lowering, but clang specifies `long double` as `"
+                 << SemName << "`\n";
+    abort();
```
This is reinventing reportFatalInternalError. But can we just fix the clang field to be set from the triple in the first place?

arsenm commented on llvm/lib/IR/RuntimeLibcalls.cpp:
```diff
@@ -62,59 +62,60 @@ static void setARMLibcallNames(RuntimeLibcallsInfo &Info, const Triple &TT,
 }
 
+/// Set all libm libcalls for _Float128 to `long double` (`*l`) symbols.
 static void setLongDoubleIsF128Libm(RuntimeLibcallsInfo &Info,
```
Hold off on this, you're jumping in the middle of my partially merged stack. #148575 removes this

tgross35 committed:
[IR] Change `fp128` lowering to use `f128` functions by default

LLVM currently emits calls to `*l` (`long double`) libm symbols for
`fp128` intrinsics. This works on platforms where `long double` and
`_Float128` are the same type, but is incorrect on many platforms.

Change RuntimeLibcalls such that `*f128` libcalls are used by default,
which is always safe and correct but may not be available. On platforms
where it is likely that `sqrtf128` and similar are not available, keep
the current behavior of lowering to `*l` symbols if `long double` is
`binary128`.

The logic for whether f128 is `long double` is based on the platforms in
Clang that set `LongDoubleFormat` to `llvm::APFloat::IEEEquad`.

Fixes https://github.com/llvm/llvm-project/issues/44744
